sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala

```diff
@@ -3621,6 +3621,40 @@ class DataFrameFunctionsSuite extends QueryTest with SharedSparkSession {
       df.select(map(map_entries($"m"), lit(1))),
       Row(Map(Seq(Row(1, "a")) -> 1)))
   }
+
+  test("SPARK-33391: element_at ArrayIndexOutOfBoundsException") {
```

Comment:
Let's not mix unit test and end-to-end test. Let's add UT in `CollectionExpressionsSuite`(for ElementAt)  and `StringExpressionsSuite` (for Elt).

For end-to-end test,  we can put them in `DataFrameSuite` for all the three functions/operators.